internal/cache/metadata/type_cache.go

```diff
@@ -0,0 +1,182 @@
+// Copyright 2015 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metadata
+
+import (
+	"fmt"
+	"math"
+	"time"
+	unsafe "unsafe"
+
+	"github.com/googlecloudplatform/gcsfuse/internal/cache/lru"
+	"github.com/googlecloudplatform/gcsfuse/internal/logger"
+	"github.com/googlecloudplatform/gcsfuse/internal/util"
+)
+
+type Type int
+
+const (
+	UnknownType     Type = 0
+	SymlinkType     Type = 1
+	RegularFileType Type = 2
+	ExplicitDirType Type = 3
+	ImplicitDirType Type = 4
+	NonexistentType Type = 5
+)
+
+func (t Type) String() string {
+	switch t {
+	case UnknownType:
+		return "UnknownType"
+	case SymlinkType:
+		return "SymlinkType"
+	case RegularFileType:
+		return "RegularFileType"
+	case ExplicitDirType:
+		return "ExplicitDirType"
+	case ImplicitDirType:
+		return "ImplicitDirType"
+	case NonexistentType:
+		return "NonexistentType"
+	}
+
+	return "Invalid value of Type"
+}
+
+// TypeCache is a (name -> Type) map.
+// It maintains TTL for each entry for supporting
+// TTL-based expiration.
+// Sample usage:
+//
+//	tc := NewTypeCache(size, ttl)
+//	tc.Insert(time.Now(), "file", RegularFileType)
+//	tc.Insert(time.Now(), "dir", ExplicitDirType)
+//	tc.Get(time.Now(),"file") -> RegularFileType
+//	tc.Get(time.Now(),"dir") -> ExplicitDirType
+//	tc.Get(time.Now()+ttl+1ns, "file") -> internally tc.Erase("file") -> UnknownType
+//	tc.Erase("dir")
+//	tc.Get(time.Now(),"dir") -> UnknownType
+type TypeCache interface {
+	// Insert inserts the given entry (name -> type)
+	// with the entry-expiration at now+ttl.
+	Insert(now time.Time, name string, it Type)
+	// Erase removes the entry with the given name.
+	Erase(name string)
+	// Get returns the entry with given name, and also
+	// records this entry as latest accessed in the cache.
+	// If now > expiration, then entry is removed from cache, and
+	// UnknownType is returned.
+	// If entry doesn't exist in the cache, then
+	// UnknownType is returned.
+	Get(now time.Time, name string) Type
+}
+
+type cacheEntry struct {
+	expiry    time.Time
+	inodeType Type
+}
+
+func (ce cacheEntry) Size() uint64 {
+	return uint64(unsafe.Sizeof(ce))
```

Comment:
Added a line for it.

Reply:
I added in the description of entries in `type typeCache struct` where the comment was showing up earlier. May be the code moved since the comment. I've moved it to this function now.